### PR TITLE
Validator integration

### PR DESCRIPTION
**Pull request by konrad0960**

### CODE REVIEW

File: neurons/rewarder.py [hunk 1]
```diff
@@ -16,19 +16,20 @@ def __init__(self, rewarder_config: RewarderConfig):
...
         else:
             days_as_leader = 0
             self.competition_leader_mapping[competition_id] = CompetitionLeader(hotkey=hotkey,
-                                                                           leader_since=datetime.now())
+                                                                           leader_since=datetime.now(timezone.utc))
+            return
         
         # Score degradation starts on 3rd week of leadership 
+        base_share = 1/len(self.competition_leader_mapping)
         if days_as_leader > 14:
```
Comment (wjurkowlaniec) [inline neurons/rewarder.py]: make a variable, some CONST or so

File: neurons/rewarder.py [hunk 2]
```diff
@@ -37,24 +38,32 @@ def get_scores_and_reduction(self, competition_id: str, hotkey: str) -> tuple[fl
             return final_share, reduced_share
         return base_share, 0
     
-    def update_scores(self):
-        num_competitions = len(self.competition_leader_mapping)
+    def update_scores(self, new_winner_hotkey: str, new_winner_comp_id: str) -> RewarderConfig:
```
Comment (wjurkowlaniec) [inline neurons/rewarder.py]: RewarderConfig -> MinerRewards

File: neurons/rewarder.py [hunk 3]
```diff
@@ -16,19 +16,20 @@ def __init__(self, rewarder_config: RewarderConfig):
         self.competition_leader_mapping = rewarder_config.competitionID_to_leader_hotkey_map
         self.scores = rewarder_config.hotkey_to_score_map
 
-    def get_scores_and_reduction(self, competition_id: str, hotkey: str) -> tuple[float, float]:
-        num_competitions = len(self.competition_leader_mapping)
-        base_share = 1/num_competitions
-
+    def get_score_and_reduction(self, competition_id: str, hotkey: str) -> tuple[float, float]:
```
Comment (wjurkowlaniec) [inline neurons/rewarder.py]: get_miner_score

File: neurons/rewarder.py [hunk 4]
```diff
@@ -37,24 +38,32 @@ def get_scores_and_reduction(self, competition_id: str, hotkey: str) -> tuple[fl
...
+        self.get_score_and_reduction(new_winner_comp_id, new_winner_hotkey)
 
+        num_competitions = len(self.competition_leader_mapping)
         # If there is only one competition, the winner takes it all
         if num_competitions == 1:
             competition_id = next(iter(self.competition_leader_mapping))
             hotkey = self.competition_leader_mapping[competition_id].hotkey
             self.scores[hotkey] = Score(score=1.0, reduction=0.0)
-            return
+            return RewarderConfig(competitionID_to_leader_hotkey_map=self.competition_leader_mapping, hotkey_to_score_map=self.scores)
```
Comment (wjurkowlaniec) [inline neurons/rewarder.py]: -

Comment (wjurkowlaniec) [reply neurons/rewarder.py]: instead add get state and load state

File: neurons/rewarder.py (hunk 1 above)
Comment (konrad0960) [reply neurons/rewarder.py]: done

File: neurons/rewarder.py (hunk 2 above)
Comment (konrad0960) [reply neurons/rewarder.py]: outdated, rewarder now doesnt returns anything

File: neurons/rewarder.py (hunk 3 above)
Comment (konrad0960) [reply neurons/rewarder.py]: Done

File: neurons/rewarder.py (hunk 4 above)
Comment (konrad0960) [reply neurons/rewarder.py]: outdated, rewarder doesnt return anything, but the overwriting of self.scores and managing local store is done on valiator level

File: neurons/competition_runner_test.py [hunk 5]
```diff
@@ -43,6 +43,7 @@ def run_all_competitions(path_config: str, hotkeys: List[str], competitions_cfg:
 def config_for_scheduler() -> Dict[str, CompetitionManager]:
     """Returns CompetitionManager instances arranged by competition time"""
     time_arranged_competitions = {}
+    main_competitions_cfg = json.load(open("neurons/competition_config.json", "r"))
```
Comment (wjurkowlaniec) [inline neurons/competition_runner_test.py]: config path should be in args, with this as default value

File: neurons/rewarder.py [hunk 6]
```diff
@@ -11,35 +11,43 @@ class RewarderConfig(BaseModel):
     competitionID_to_leader_hotkey_map: dict[str, CompetitionLeader] # competition_id -> CompetitionLeader
     hotkey_to_score_map: dict[str, Score] # hotkey -> Score
 
+NON_REDUCTION_PERIOD = 30
```
Comment (wjurkowlaniec) [inline neurons/rewarder.py]: REWARD_REDUCTION_START_DAY = 30
REWARD_REDUCTION_STEP = 0.1
REWARD_REDUCTION_STEP_DAYS = 7
REWARD_REDUCTION_MAX = 0.9

File: neurons/rewarder_test.py [hunk 7]
```diff
@@ -73,10 +72,10 @@ def test_update_scores_multiple_competitors_no_reduction():
 def test_update_scores_multiple_competitors_with_some_reduced_shares():
     # Set up initial data for multiple competitors
     competition_leaders = {
-        "competition1": CompetitionLeader(hotkey="competitor1", leader_since=datetime.now() - timedelta(days=14 + 3 * 7)),
-        "competition2": CompetitionLeader(hotkey="competitor2", leader_since=datetime.now() - timedelta(days=14 + 6 * 7)),
-        "competition3": CompetitionLeader(hotkey="competitor3", leader_since=datetime.now() - timedelta(days=14)),
-        "competition4": CompetitionLeader(hotkey="competitor4", leader_since=datetime.now() - timedelta(days=14)),
+        "competition1": CompetitionLeader(hotkey="competitor1", leader_since=datetime.now() - timedelta(days=30 + 3 * 7)),
```
Comment (wjurkowlaniec) [inline neurons/rewarder_test.py]: you can change "competitior1" for a name "competitior_30_percent_reduction"

Comment (wjurkowlaniec) [reply neurons/rewarder_test.py]: etc

File: neurons/validator.py [hunk 8]
```diff
@@ -44,12 +46,68 @@ class Validator(BaseValidatorNeuron):
...
+        bt.logging.info("Running Competition Loop.")
+        while True:
+            competition_result = await run_competitions_tick(scheduler_config)
+            bt.logging.debug(f"Competition result: {competition_result}")
+            if competition_result:
+                winning_evaluation_hotkey, competition_id = competition_result
+
+                # update the scores
+                await self.rewarder.update_scores(winning_evaluation_hotkey, competition_id)
+                self.rewarder_config = RewarderConfig(self.rewarder.competition_leader_mapping, self.rewarder.scores)
```
Comment (wjurkowlaniec) [inline neurons/validator.py]: rewarderconfig doesn't fit here

Comment (konrad0960) [reply neurons/validator.py]: we store rewarder config locally and it needs to be updated after update_scores is run
isnt updating it right away the most safe way?

File: neurons/rewarder_test.py (hunk 7 above)
Comment (konrad0960) [reply neurons/rewarder_test.py]: this is mobbing

File: scripts/start_validator.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        format="%(asctime)s %(levelname)s %(message)s",
+        handlers=[logging.StreamHandler(sys.stdout)],
+    )
+
+    parser = argparse.ArgumentParser(
+        description="Automatically update and restart the validator process when a new version is released.",
+        epilog="Example usage: python start_validator.py --pm2_name 'net9vali' --wallet_name 'wallet1' --wallet_hotkey 'key123'",
+    )
+
+    parser.add_argument(
```
Comment (wjurkowlaniec) [inline scripts/start_validator.py:195]: remove args, and put in documentation that  you need to edit pm2 file.

Also move pm2 file to config directory

File: neurons/competition_runner.py [hunk 10]
```diff
@@ -1,64 +1,155 @@
...
-from types import SimpleNamespace
 from datetime import datetime, timezone, timedelta
 import bittensor as bt
 from typing import List, Tuple, Dict
+from cancer_ai.validator.rewarder import Rewarder, WinnersMapping, CompetitionLeader
 
 # from cancer_ai.utils.config import config
 
+# TODO MOVE SOMEWHERE
+main_competitions_cfg = json.load(open("neurons/competition_config.json", "r"))
```
Comment (wjurkowlaniec) [inline neurons/competition_runner.py:14]: move config file to `config/competition_config.json`

**Pull request merged**